File: contents/contents/quantum_systems/code/julia/energy.jl
# We are calculating the energy to check <Psi|H|Psi>
function calculate_energy(wfc, H_k, H_r, dx)
    # Creating momentum and conjugate wavefunctions
    wfc_k = fft(wfc)
    wfc_c = conj(wfc)

    # Finding the momentum and real-space energy terms
    energy_k = wfc_c.*ifft((H_k) .* wfc_k)
    energy_r = wfc_c.* H_r .* wfc

    # Integrating over all space
    energy_final = 0
    for i = 1:length(energy_k)
        energy_final += real(energy_k[i] + energy_r[i])
    end 

    return energy_final*dx
end

File: contents/contents/quantum_systems/code/python/energy.py
import numpy as np


def calculate_energy(wfc, H_k, H_r, dx):
    """Calculate the energy <Psi|H|Psi>."""
    # Creating momentum conjugate wavefunctions
    wfc_k = np.fft.fft(wfc)
    wfc_c = np.conj(wfc)

    # Finding the momentum and real-space energy terms
    energy_k = 0.5 * wfc_c * np.fft.ifft((H_k ** 2) * wfc_k)
    energy_r = wfc_c * H_r * wfc

    # Integrating over all space
    energy_final = sum(energy_k + energy_r).real

    return energy_final * dx
